##  Search
Name: Jonathan  Lopez

Team Mate: Andrew Reigle

### The following code snippet uses a dictionary of dictionaries to encode the Romania roadmap and demonstrates how to get the elements.

In [1]:
roadmap = dict(
    Arad=dict(Zerind=75, Sibiu=140, Timisoara=118),
    Bucharest=dict(Urziceni=85, Pitesti=101, Giurgiu=90, Fagaras=211),
    Craiova=dict(Drobeta=120, Rimnicu=146, Pitesti=138),
    Drobeta=dict(Craiova=120, Mehadia=75),
    Eforie=dict(Hirsova=86),
    Fagaras=dict(Bucharest=211, Sibiu=99),
    Giurgiu=dict(Bucharest=90),
    Hirsova=dict(Eforie=86, Urziceni=98),
    Iasi=dict(Vaslui=92, Neamt=87),
    Lugoj=dict(Timisoara=111, Mehadia=70),
    Mehadia=dict(Lugoj=70, Drobeta=75),
    Neamt=dict(Iasi=87),
    Oradea=dict(Zerind=71, Sibiu=151),
    Pitesti=dict(Bucharest=101, Rimnicu=97, Craiova=138),
    Rimnicu=dict(Craiova=146, Sibiu=80, Pitesti=97),
    Sibiu=dict(Arad=140, Fagaras=99, Oradea=151, Rimnicu=80),
    Timisoara=dict(Arad=118,Lugoj=111),
    Vaslui=dict(Iasi=92, Urziceni=98),
    Urziceni=dict(Vaslui=142, Bucharest=85, Hirsova=98),
    Zerind=dict(Arad=75, Oradea=71))

# Demonstration of how to use this data structure
#  by printing the map elements and performing a sanity check -
# for city, neighbors in roadmap.items():
#   print(city, neighbors)
#   for next_city, distance in neighbors.items():
#     print("  The distance from %s to %s is %d" % (city, next_city, distance))
#     if next_city not in roadmap:
#       print(" ----- ERROR! %s is not in the city node list -----" % next_city)


**Your code starts below.  Please organized them in multiple cells with a formatted text cell before each code cell to explain the purpose of the code cell**

The BFS implementation explores all nodes at the current depth level before moving to the next level and uses a FIFO queue for the management of the frontier. Takes in arguments for starting city, goal city  and dictionary of the roadmap, returning the path, total cost and expanded nodes.

In [7]:
#BFS Implementation
from collections import deque

def bfs(start, goal, roadmap):
    #initialize frontier with the start node
    frontier = deque([{'city': start, 'parent': None, 'cost': 0}])
    reached = {} # tracks reached nodes and their associated costs
    expanded = [] # tracks expanded nodes for performance comparison

    while frontier:
        current_node = frontier.popleft()
        current_city = current_node['city']

        #skip if already reached with a lower cost
        if current_city in reached:
            continue
        reached[current_city] = current_node['cost']
        # expanded.append(current_city) *** commented out to change adding bucharest to nodes expanded

        # Check if goal is reached, then reconstruct path
        if current_city == goal:
            path = []
            node = current_node
            while node:
                path.append(node['city'])
                node = node['parent']
            path.reverse()
            return {
                'path': path,
                'cost': current_node['cost'],
                'expanded': expanded
            }

         # Now we add the current node since it's not the goal.
        expanded.append(current_city)

        # Expand neighbors and add to frontier
        for neighbor in roadmap[current_city]:
            new_cost = current_node['cost'] + roadmap[current_city][neighbor]

            # ***without the condition to check reached, we add duplicate paths(parent nodes)
            # frontier.append({'city': neighbor, 'parent': current_node, 'cost': new_cost})

            #***added this condition to ensure only new or cheaper paths are added to frontier to reduce redunancy
            if neighbor not in reached or new_cost < reached.get(neighbor, float('inf')):
                frontier.append({'city': neighbor, 'parent': current_node, 'cost': new_cost})

        # Print step details
        parent_str = current_node['parent']['city'] if current_node['parent'] else None
        current_str = f"{parent_str}->{current_city}({current_node['cost']})" if parent_str else f"{current_city}(0)"
        print(f"BFS Node expanded: {current_str}")
        frontier_str = [f"{node['parent']['city']}->{node['city']}({node['cost']})" if node['parent'] else f"{node['city']}(0)" for node in frontier]
        print(f" BFS Frontier: {frontier_str}\n")

    return None

# Test BFS
bfs_result = bfs('Arad', 'Bucharest', roadmap)
print("BFS Path:", bfs_result['path'])
print("Total Cost:", bfs_result['cost'])
print("Nodes Expanded:", bfs_result['expanded'])

BFS Node expanded: Arad(0)
 BFS Frontier: ['Arad->Zerind(75)', 'Arad->Sibiu(140)', 'Arad->Timisoara(118)']

BFS Node expanded: Arad->Zerind(75)
 BFS Frontier: ['Arad->Sibiu(140)', 'Arad->Timisoara(118)', 'Zerind->Oradea(146)']

BFS Node expanded: Arad->Sibiu(140)
 BFS Frontier: ['Arad->Timisoara(118)', 'Zerind->Oradea(146)', 'Sibiu->Fagaras(239)', 'Sibiu->Oradea(291)', 'Sibiu->Rimnicu(220)']

BFS Node expanded: Arad->Timisoara(118)
 BFS Frontier: ['Zerind->Oradea(146)', 'Sibiu->Fagaras(239)', 'Sibiu->Oradea(291)', 'Sibiu->Rimnicu(220)', 'Timisoara->Lugoj(229)']

BFS Node expanded: Zerind->Oradea(146)
 BFS Frontier: ['Sibiu->Fagaras(239)', 'Sibiu->Oradea(291)', 'Sibiu->Rimnicu(220)', 'Timisoara->Lugoj(229)']

BFS Node expanded: Sibiu->Fagaras(239)
 BFS Frontier: ['Sibiu->Oradea(291)', 'Sibiu->Rimnicu(220)', 'Timisoara->Lugoj(229)', 'Fagaras->Bucharest(450)']

BFS Node expanded: Sibiu->Rimnicu(220)
 BFS Frontier: ['Timisoara->Lugoj(229)', 'Fagaras->Bucharest(450)', 'Rimnicu->Craiova(366)

The Depth-first search implementation explores the deepest node first using a LIFO stack, so it does not guarantee the shortest path. Takes in the same arguments as BFS(starting city, goal strings and roadmap dictionary), returning the reached and expaned nodes.

In [8]:
#DFS Implementation

def dfs(start, goal, roadmap):
    frontier = [{'city': start, 'parent': None, 'cost': 0}]
    reached = {}  # Track {city: best_cost}
    expanded = []

    while frontier:
        current_node = frontier.pop()
        current_city = current_node['city']

        # Skip if already reached with a lower or equal cost
        if current_city in reached and current_node['cost'] >= reached[current_city]:
            continue
        reached[current_city] = current_node['cost']  # Update best cost
        # expanded.append(current_city) *** also commented out to prevent adding goal node to total nodes expanded

        # Check if goal is reached
        if current_city == goal:
            path = []
            node = current_node
            while node:
                path.append(node['city'])
                node = node['parent']
            path.reverse()
            return {
                'path': path,
                'cost': current_node['cost'],
                'expanded': expanded
            }

        # Now add the node to expanded only if it's not the goal.
        expanded.append(current_city)

        # Expand neighbors in reverse order (for LIFO)
        neighbors = list(roadmap[current_city].keys())[::-1]
        for neighbor in neighbors:
            new_cost = current_node['cost'] + roadmap[current_city][neighbor]
            # Add to frontier ONLY if neighbor not in reached or new_cost < existing cost
            if neighbor not in reached or new_cost < reached.get(neighbor, float('inf')):
                frontier.append({'city': neighbor, 'parent': current_node, 'cost': new_cost})

        # Print step
        parent_str = current_node['parent']['city'] if current_node['parent'] else None
        current_str = f"{parent_str}->{current_city}({current_node['cost']})" if parent_str else f"{current_city}(0)"
        print(f"DFS Node expanded: {current_str}")
        frontier_str = [f"{node['parent']['city']}->{node['city']}({node['cost']})" if node['parent'] else f"{node['city']}(0)" for node in frontier]
        print(f"DFS Frontier: {frontier_str}\n")

    return None


# Test DFS
dfs_result = dfs('Arad', 'Bucharest', roadmap)
print("DFS Path:", dfs_result['path'])
print("Total Cost:", dfs_result['cost'])
print("Nodes Expanded:", dfs_result['expanded'])



DFS Node expanded: Arad(0)
DFS Frontier: ['Arad->Timisoara(118)', 'Arad->Sibiu(140)', 'Arad->Zerind(75)']

DFS Node expanded: Arad->Zerind(75)
DFS Frontier: ['Arad->Timisoara(118)', 'Arad->Sibiu(140)', 'Zerind->Oradea(146)']

DFS Node expanded: Zerind->Oradea(146)
DFS Frontier: ['Arad->Timisoara(118)', 'Arad->Sibiu(140)', 'Oradea->Sibiu(297)']

DFS Node expanded: Oradea->Sibiu(297)
DFS Frontier: ['Arad->Timisoara(118)', 'Arad->Sibiu(140)', 'Sibiu->Rimnicu(377)', 'Sibiu->Fagaras(396)']

DFS Node expanded: Sibiu->Fagaras(396)
DFS Frontier: ['Arad->Timisoara(118)', 'Arad->Sibiu(140)', 'Sibiu->Rimnicu(377)', 'Fagaras->Bucharest(607)']

DFS Path: ['Arad', 'Zerind', 'Oradea', 'Sibiu', 'Fagaras', 'Bucharest']
Total Cost: 607
Nodes Expanded: ['Arad', 'Zerind', 'Oradea', 'Sibiu', 'Fagaras']


A* search is supposed to combine path cost g(n) and heuristic h(n) in order to prioritize nodes. This implementation uses a priority queue sorted by the sum of both path and heuristic costs. This implimentation required creating a heuristic dictionary of the straight line distances to Bucharest.

In [9]:
#A Search* Implementation
import heapq # used to impliment a priority queue

# Heuristic values (straight-line distance to Bucharest)
heuristic = {
    'Arad': 366, 'Bucharest': 0, 'Craiova': 160, 'Drobeta': 242, 'Eforie': 161,
    'Fagaras': 176, 'Giurgiu': 77, 'Hirsova': 151, 'Iasi': 226, 'Lugoj': 244,
    'Mehadia': 241, 'Neamt': 234, 'Oradea': 380, 'Pitesti': 100, 'Rimnicu': 193,
    'Sibiu': 253, 'Timisoara': 329, 'Urziceni': 80, 'Vaslui': 199, 'Zerind': 374
}
#Priority queue
def a_star(start, goal, roadmap, heuristic):
    frontier = []

    ## (estimated_total_cost = cost_so_far + heuristic, cost_so_far, current_city, parent pointer)
    heapq.heappush(frontier, (0 + heuristic[start], 0, start, None))
    reached = {}
    expanded = []

    #continue searching until there are no more nodes in the frontier
    while frontier:
        #pop node with the smallest estimated total cost
        current_total, current_cost, current_city, parent = heapq.heappop(frontier)

        #if current city is goal, reconstruct the path by following parent pointers
        if current_city == goal:
            path = []
            while parent:
                path.append(current_city)
                current_city, parent = parent
            path.append(start)
            path.reverse() # reverese the path to show it from start
            return {
                'path': path,
                'cost': current_cost,
                'expanded': expanded
            }

        #if we already reached this city at lower cost, skip further processing
        if current_city in reached and current_cost >= reached[current_city]:
            continue
        reached[current_city] = current_cost
        expanded.append(current_city)

        # Expand neighbors of current city
        for neighbor in roadmap[current_city]:
            new_cost = current_cost + roadmap[current_city][neighbor]
            new_total = new_cost + heuristic[neighbor]

            #if this is first time reaching or if path is cheaper, then add to frontier
            if neighbor not in reached or new_cost < reached.get(neighbor, float('inf')):
                heapq.heappush(frontier, (new_total, new_cost, neighbor, (current_city, parent)))

        # Print step
        parent_str = parent[0] if parent else None
        current_str = f"{parent_str}->{current_city}({current_cost}, {current_total})" if parent_str else f"{current_city}(0, {heuristic[start]})"
        print(f" A* Node expanded: {current_str}")
        frontier_list = []
        for entry in frontier:
            total, cost, city, prnt = entry
            prnt_city = prnt[0] if prnt else None
            frontier_list.append(f"{prnt_city}->{city}({cost}, {total})" if prnt_city else f"{city}({cost}, {total})")
        print(f"A* Frontier: {frontier_list}\n")

    return None

# Test A*
a_star_result = a_star('Arad', 'Bucharest', roadmap, heuristic)
print("A* Path:", a_star_result['path'])
print("Total Cost:", a_star_result['cost'])
print("Nodes Expanded:", a_star_result['expanded'])

 A* Node expanded: Arad(0, 366)
A* Frontier: ['Arad->Sibiu(140, 393)', 'Arad->Zerind(75, 449)', 'Arad->Timisoara(118, 447)']

 A* Node expanded: Arad->Sibiu(140, 393)
A* Frontier: ['Sibiu->Rimnicu(220, 413)', 'Sibiu->Fagaras(239, 415)', 'Arad->Timisoara(118, 447)', 'Sibiu->Oradea(291, 671)', 'Arad->Zerind(75, 449)']

 A* Node expanded: Sibiu->Rimnicu(220, 413)
A* Frontier: ['Sibiu->Fagaras(239, 415)', 'Arad->Zerind(75, 449)', 'Rimnicu->Pitesti(317, 417)', 'Sibiu->Oradea(291, 671)', 'Rimnicu->Craiova(366, 526)', 'Arad->Timisoara(118, 447)']

 A* Node expanded: Sibiu->Fagaras(239, 415)
A* Frontier: ['Rimnicu->Pitesti(317, 417)', 'Arad->Zerind(75, 449)', 'Arad->Timisoara(118, 447)', 'Sibiu->Oradea(291, 671)', 'Rimnicu->Craiova(366, 526)', 'Fagaras->Bucharest(450, 450)']

 A* Node expanded: Rimnicu->Pitesti(317, 417)
A* Frontier: ['Pitesti->Bucharest(418, 418)', 'Arad->Zerind(75, 449)', 'Arad->Timisoara(118, 447)', 'Sibiu->Oradea(291, 671)', 'Rimnicu->Craiova(366, 526)', 'Fagaras->Buchares

**Extra Credit - algorithm analysis**
Runs BFS, DFS and A* and displays solution path, cost, time taken and nodes expanded for comparison.

In [33]:

import time

#Measure BFS Performance
start_time = time.perf_counter()
bfs_result = bfs('Arad', 'Bucharest', roadmap)
end_time = time.perf_counter()
bfs_time = end_time - start_time

#Measure DFS Performance
start_time = time.perf_counter()
dfs_result = dfs('Arad', 'Bucharest', roadmap)
end_time = time.perf_counter()
dfs_time = end_time - start_time

# Measure A* Performance
start_time = time.perf_counter()
astar_result = a_star('Arad', 'Bucharest', roadmap, heuristic)
end_time = time.perf_counter()
astar_time = end_time - start_time

#  Print performance results
print("=== BFS ===")
print("Path:", bfs_result['path'])
print("Cost:", bfs_result['cost'])
print("Nodes Expanded:", len(bfs_result['expanded']))
print("Time: {:.6f} seconds".format(bfs_time))

print("\n=== DFS ===")
print("Path:", dfs_result['path'])
print("Cost:", dfs_result['cost'])
print("Nodes Expanded:", len(dfs_result['expanded']))
print("Time: {:.6f} seconds".format(dfs_time))

print("\n=== A* ===")
print("Path:", astar_result['path'])
print("Cost:", astar_result['cost'])
print("Nodes Expanded:", len(astar_result['expanded']))
print("Time: {:.6f} seconds".format(astar_time))

#Print Comparison Table
print("\n=== Performance Comparison ===")
print("Algorithm | Path Cost | Nodes Expanded |   Time (s)")
print("---------------------------------------------------")
print("BFS       | {:9} | {:14} | {:.6f}".format(bfs_result['cost'], len(bfs_result['expanded']), bfs_time))
print("DFS       | {:9} | {:14} | {:.6f}".format(dfs_result['cost'], len(dfs_result['expanded']), dfs_time))
print("A*        | {:9} | {:14} | {:.6f}".format(astar_result['cost'], len(astar_result['expanded']), astar_time))



BFS Node expanded: Arad(0)
 BFS Frontier: ['Arad->Zerind(75)', 'Arad->Sibiu(140)', 'Arad->Timisoara(118)']

BFS Node expanded: Arad->Zerind(75)
 BFS Frontier: ['Arad->Sibiu(140)', 'Arad->Timisoara(118)', 'Zerind->Oradea(146)']

BFS Node expanded: Arad->Sibiu(140)
 BFS Frontier: ['Arad->Timisoara(118)', 'Zerind->Oradea(146)', 'Sibiu->Fagaras(239)', 'Sibiu->Oradea(291)', 'Sibiu->Rimnicu(220)']

BFS Node expanded: Arad->Timisoara(118)
 BFS Frontier: ['Zerind->Oradea(146)', 'Sibiu->Fagaras(239)', 'Sibiu->Oradea(291)', 'Sibiu->Rimnicu(220)', 'Timisoara->Lugoj(229)']

BFS Node expanded: Zerind->Oradea(146)
 BFS Frontier: ['Sibiu->Fagaras(239)', 'Sibiu->Oradea(291)', 'Sibiu->Rimnicu(220)', 'Timisoara->Lugoj(229)']

BFS Node expanded: Sibiu->Fagaras(239)
 BFS Frontier: ['Sibiu->Oradea(291)', 'Sibiu->Rimnicu(220)', 'Timisoara->Lugoj(229)', 'Fagaras->Bucharest(450)']

BFS Node expanded: Sibiu->Rimnicu(220)
 BFS Frontier: ['Timisoara->Lugoj(229)', 'Fagaras->Bucharest(450)', 'Rimnicu->Craiova(366)

Observations: Based on our results, BFS expanded the most nodes and returned a path with a cost of 450, but in a weighted graph it only guarantees the fewest-edge path rather than the optimal cost. DFS, while expanding fewer nodes in this run, ended up with a significantly higher-cost path (607) due to its deep backtracking nature. In contrast, A* search efficiently incorporating heuristic values and identified the optimal path with a cost of 418, while expanding only 5 nodes.
This illustrates the trade-offs among the search strategies: BFS is exhaustive but not cost-optimal in weighted scenarios, DFS can be fast yet suboptimal, and A* effectively balances exploration and cost optimization through its heuristic guidance.